File: plugins/lcfarm_flutter_umeng/ios/Classes/LcfarmFlutterUmengPlugin.m
#import "LcfarmFlutterUmengPlugin.h"
#import <UMCommon/MobClick.h>
#import <UMCommon/UMCommon.h>

@implementation LcfarmFlutterUmengPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"lcfarm_flutter_umeng"
                                     binaryMessenger:[registrar messenger]];
    LcfarmFlutterUmengPlugin* instance = [[LcfarmFlutterUmengPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    //初始化
    if ([@"init" isEqualToString:call.method]) {
        [self init:call result:result];
        result(nil);
    }
    //事件埋点
    else if ([@"event" isEqualToString:call.method]) {
        [self event:call result:result];
        result(nil);
    }
    //统计页面时间-开始
    else if ([@"beginLogPageView" isEqualToString:call.method]){
        [self beginLogPageView:call result:result];
        result(nil);
    }
    //统计页面时间-结束
    else if ([@"endLogPageView" isEqualToString:call.method]){
        [self endLogPageView:call result:result];
        result(nil);
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)init:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    NSNumber* logEnable = call.arguments[@"logEnable"];
    if (logEnable) {
        [UMConfigure setLogEnabled:[logEnable boolValue]];
    }
    NSNumber* encrypt = call.arguments[@"encrypt"];
    if (encrypt) {
        [UMConfigure setEncryptEnabled:[encrypt boolValue]];
    }
    
    NSString* channel = call.arguments[@"channel"] ?: @"App Store";
    
    NSString* appKey = call.arguments[@"appKey"] ?: @"";
    
    [UMConfigure initWithAppkey:appKey channel:channel];
    
}

- (void)event:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    if (call.arguments[@"label"]){
        [MobClick event:call.arguments[@"eventId"] label:call.arguments[@"label"]];
    }else{
        [MobClick event:call.arguments[@"eventId"]];
    }
}

- (void)beginLogPageView:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    [MobClick beginLogPageView:call.arguments[@"pageName"] ?: @""];
}

- (void)endLogPageView:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    [MobClick endLogPageView:call.arguments[@"pageName"] ?: @""];
    
}

@end

File: plugins/lcfarm_flutter_umeng/android/src/main/java/com/nongfadai/lcfarm_flutter_umeng/LcfarmFlutterUmengPlugin.java
package com.nongfadai.lcfarm_flutter_umeng;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.text.TextUtils;

import com.umeng.analytics.MobclickAgent;
import com.umeng.commonsdk.UMConfigure;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** LcfarmFlutterUmengPlugin */
public class LcfarmFlutterUmengPlugin implements MethodCallHandler {

    private Activity activity;

    /** Plugin registration. */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "lcfarm_flutter_umeng");
        channel.setMethodCallHandler(new LcfarmFlutterUmengPlugin(registrar.activity()));
    }

    private LcfarmFlutterUmengPlugin(Activity activity) {
        this.activity = activity;
    }


    public static String getChannel(Context context) {
        try {
            ApplicationInfo appInfo = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
            String channel = appInfo.metaData.getString("UMENG_CHANNEL");
            return channel;
        } catch (PackageManager.NameNotFoundException e) {

        }
        return null;
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {

        if (call.method.equals("init")){
            init(call,result);
        }else if (call.method.equals("event")){
            event(call, result);
        }else if (call.method.equals("beginLogPageView")){
            beginLogPageView(call, result);
        }else if (call.method.equals("endLogPageView")){
            endLogPageView(call, result);
        }else if (call.method.equals("onResume")){
            onResumeActivity(call, result);
        }else if (call.method.equals("onPause")){
            onPauseActivity(call, result);
        }

        else {
            result.notImplemented();
        }
    }


    public void init(MethodCall call, Result result){

        boolean logEnable = false;
        if(call.hasArgument("logEnable")){
            logEnable = (boolean)call.argument("logEnable");
        }

        UMConfigure.setLogEnabled(logEnable);

        if(call.hasArgument("channel")){
            String channel = (String) call.argument("channel");
            UMConfigure.init(activity, (String) call.argument("appKey"), channel, UMConfigure.DEVICE_TYPE_PHONE, null);
        }else {

            String channel = getChannel(activity);
            UMConfigure.init(activity, (String) call.argument("appKey"), channel, UMConfigure.DEVICE_TYPE_PHONE, null);
        }

        boolean encrypt = false;

        if(call.hasArgument("encrypt")){
            encrypt = (boolean)call.argument("encrypt");
        }

        UMConfigure.setEncryptEnabled(encrypt);

        MobclickAgent.openActivityDurationTrack(false);

        result.success(true);

    }

    public  void event(MethodCall call, Result result) {

        if (call.hasArgument("label")){
            String label = (String) call.argument("label");
            MobclickAgent.onEvent(activity,(String) call.argument("eventId"),label);

        }else {
            MobclickAgent.onEvent(activity, (String) call.argument("eventId"));
        }

        result.success(true);
    }

    public  void beginLogPageView(MethodCall call, Result result) {

        MobclickAgent.onPageStart((String) call.argument("pageName"));

        result.success(true);

    }

    public  void endLogPageView(MethodCall call, Result result) {

        MobclickAgent.onPageEnd((String) call.argument("pageName"));

        result.success(true);

    }

    public  void onResumeActivity(MethodCall call, Result result) {

        MobclickAgent.onPause(activity);

        result.success(true);

    }

    public  void onPauseActivity(MethodCall call, Result result) {

        MobclickAgent.onPause(activity);

        result.success(true);

    }

}
